src/runtime/string.cpp

```diff
@@ -672,8 +672,70 @@ static PRIMFN(prim_shell_str) {
   RETURN(String::alloc(runtime.heap, shell_escape(str->c_str())));
 }
 
+static PRIMTYPE(type_substr) {
+  TypeVar result;
+  Data::typeResult.clone(result);
+  result[0].unify(Data::typeString);
+  result[1].unify(Data::typeString);
+  return args.size() == 3 &&
+    args[0]->unify(Data::typeString) &&
+    args[1]->unify(Data::typeInteger) &&
+    args[2]->unify(Data::typeInteger) &&
+    out->unify(result);
+}
+
+// Create a 'Result String String' value on the heap.
+Value *allocate_result(Heap &h, bool is_pass, const std::string &msg) {
+    h.reserve(reserve_result() + String::reserve(msg.size()));
+    return claim_result(h, is_pass, String::claim(h, msg));
+}
+
+// This operates on utf8 runes rather than bytes, to match prim_explode.
+static PRIMFN(prim_substr) {
+  EXPECT(3);
+  STRING(input_str, 0);
+  INTEGER_MPZ(pos_arg, 1);
+  INTEGER_MPZ(len_arg, 2);
+
+  int start_pos = mpz_get_si(pos_arg);
+  int length = mpz_get_si(len_arg);
```

Comment:
Erg. `int` is only 32 bits large. Strings in wake can be larger than 4GiB. You should use either `long` or `size_t`.

Reply:
Hmm. On windows, even `long` is 32 bits.

Reply:
GMP `mpz_get_si` returns `long`, so I guess that's the best you can hope for.
However, you definitely also need to check `mpz_fits_slong_p` in case the integer won't fit.

Reply:
Fixed by adding explicit bounds checks before conversion to an unsigned long.